Unit tests fail randomly: test_move_to_zero

Unit tests are failing in a set of tests.  The tests are not rock-solid reliable.

```
        short_delay_for_EPICS_IOC_database_processing()
        assert status.done
        assert status.success
>       assert status.elapsed >= 0.5
E       assert 0.002117156982421875 >= 0.5
E        +  where 0.002117156982421875 = MoveStatus(done=True, pos=pos, elapsed=0.0, success=True, settle_time=0.0).elapsed
/home/runner/work/apstools/apstools/apstools/devices/tests/test_positioner_soft_done.py:233: AssertionError
=============================== warnings summary ===============================
apstools/utils/tests/test_listruns_class.py: 27 warnings
apstools/utils/tests/test_utils.py: 4 warnings
  /home/runner/micromamba/envs/anaconda-test-env-py-3.7/lib/python3.9/site-packages/databroker/queries.py:89: PytzUsageWarning: The zone attribute is specific to pytz's interface; please migrate to a new time zone provider. For more details on how to do so, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
    timezone = lz.zone
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED apstools/devices/tests/test_positioner_soft_done.py::test_move_to_zero
============ 1 failed, 369 passed, 31 warnings in 193.77s (0:03:13) ============
Error: Process completed with exit code 1.
```

_Originally posted by @prjemian in https://github.com/BCDA-APS/apstools/issues/667#issuecomment-1176596283_

```
        short_delay_for_EPICS_IOC_database_processing()
        assert pos.setpoint.get() == 0
>       assert round(pos.position, 2) == 0
E       AssertionError: assert 5.24 == 0
E        +  where 5.24 = round(5.24, 2)
E        +    where 5.24 = PVPositionerSoftDoneWithStop(prefix='gp:gp:', name='pos', settle_time=0.0, timeout=None, read_attrs=['readback', 'setpoint'], configuration_attrs=['done', 'tolerance', 'target'], limits=None, egu='').position
/home/runner/work/apstools/apstools/apstools/devices/tests/test_positioner_soft_done.py:241: AssertionError
=============================== warnings summary ===============================
apstools/utils/tests/test_listruns_class.py: 27 warnings
apstools/utils/tests/test_utils.py: 4 warnings
  /home/runner/micromamba/envs/anaconda-test-env-py-3.8/lib/python3.9/site-packages/databroker/queries.py:89: PytzUsageWarning: The zone attribute is specific to pytz's interface; please migrate to a new time zone provider. For more details on how to do so, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
    timezone = lz.zone
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED apstools/devices/tests/test_positioner_soft_done.py::test_move_to_zero
============ 1 failed, 369 passed, 31 warnings in 170.79s (0:02:50) ============
Error: Process completed with exit code 1.
```

Still, this code failed 2 out of 3 times in the GH Actions workflow:

```
>       assert round(status.elapsed, 1) == delay_time
E       assert 0.0 == 0.2
E        +  where 0.0 = round(0.0016665458679199219, 1)
E        +    where 0.0016665458679199219 = MoveStatus(done=True, pos=pos, elapsed=0.0, success=True, settle_time=0.0).elapsed
```

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -230,7 +230,12 @@ def test_move_to_zero(target, rbv, pos):
     # start from known position
     known_start_position = -1
     pos.setpoint.put(known_start_position)
+    short_delay_for_EPICS_IOC_database_processing()
     rbv.put(known_start_position)  # note: pos.readback is read-only
+    short_delay_for_EPICS_IOC_database_processing(1)
+    assert pos.inposition
+    assert pos.setpoint.get(use_monitor=False) == known_start_position
+    assert round(pos.readback.get(use_monitor=False), 2) == known_start_position
 
     delay_time = round(0.2 + 0.7 * random.random(), 1)
 

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -1,6 +1,3 @@
-import ophyd
-ophyd.set_cl("caproto")
-
 from ophyd import EpicsSignal
 
 from ..positioner_soft_done import PVPositionerSoftDone

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -228,10 +228,10 @@ def test_move_to_zero(rbv, pos):
 
     def make_verified_move(target, delay_time):
         # start code that will update the RBV later
-        delayed_complete(pos, rbv, delay=delay_time)  
+        delayed_complete(pos, rbv, delay=delay_time)
 
         # start the move and wait for RBV to be updated
-        status = pos.move(target)  
+        status = pos.move(target)
 
         short_delay_for_EPICS_IOC_database_processing()
 

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -261,9 +261,10 @@ def test_move_to_zero(target, rbv, pos):
     # # FIXME: Why does this return immediately SOMETIMES in GH Actions workflow?
     # # assert round(status.elapsed, 1) in (0.0, delay_time)
 
-    # # verify the readback was updated AFTER the setpoint with correct delay
-    # dt = (
-    #     pos.readback.read()["pos"]["timestamp"]
-    #     - pos.setpoint.read()["pos_setpoint"]["timestamp"]
-    # )
+    # verify the readback was updated AFTER the setpoint with correct delay
+    dt = (
+        pos.readback.read()["pos"]["timestamp"]
+        - pos.setpoint.read()["pos_setpoint"]["timestamp"]
+    )
     # assert round(dt, 1) in (0.0, delay_time), f"dt={dt}, exp={delay_time}"
+    assert round(dt, 1) == delay_time, f"dt={dt}, exp={delay_time}"

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -233,6 +233,8 @@ def test_move_to_zero(target, rbv, pos):
     short_delay_for_EPICS_IOC_database_processing()
     rbv.put(known_start_position)  # note: pos.readback is read-only
     short_delay_for_EPICS_IOC_database_processing(1)
+
+    # confirm all are ready
     assert pos.inposition
     assert pos.setpoint.get(use_monitor=False) == known_start_position
     assert round(pos.readback.get(use_monitor=False), 2) == known_start_position
@@ -257,14 +259,10 @@ def test_move_to_zero(target, rbv, pos):
 
     assert status.done
     assert status.success
-    # Are these remaining tests needed?
-    # # FIXME: Why does this return immediately SOMETIMES in GH Actions workflow?
-    # # assert round(status.elapsed, 1) in (0.0, delay_time)
 
     # verify the readback was updated AFTER the setpoint with correct delay
     dt = (
         pos.readback.read()["pos"]["timestamp"]
         - pos.setpoint.read()["pos_setpoint"]["timestamp"]
     )
-    # assert round(dt, 1) in (0.0, delay_time), f"dt={dt}, exp={delay_time}"
     assert round(dt, 1) == delay_time, f"dt={dt}, exp={delay_time}"

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -223,48 +223,42 @@ def test_move_and_stopped_early(rbv, pos):
     assert pos.inposition
 
 
-def test_move_to_zero(rbv, pos):
+@pytest.mark.parametrize("target", [0, round(2 + 5 * random.random(), 2), 0, 0, 0])
+def test_move_to_zero(target, rbv, pos):
     short_delay_for_EPICS_IOC_database_processing()
-
-    def make_verified_move(target, delay_time):
-        # start code that will update the RBV later
-        delayed_complete(pos, rbv, delay=delay_time)
-
-        # start the move and wait for RBV to be updated
-        status = pos.move(target)
-
-        short_delay_for_EPICS_IOC_database_processing()
-
-        assert status.done
-        assert status.success
-        # FIXME: Why does this return immediately SOMETIMES in GH Actions workflow?
-        assert round(status.elapsed, 1) in (0.0, delay_time)
-
-        # assert not delay_active
-        assert pos.setpoint.get() == target
-
-        # verify the readback was updated AFTER the setpoint with correct delay
-        dt = (
-            pos.readback.read()["pos"]["timestamp"]
-            - pos.setpoint.read()["pos_setpoint"]["timestamp"]
-        )
-        assert round(dt, 1) in (0.0, delay_time), f"dt={dt}, exp={delay_time}"
-
-        # note: pos.position has been failing (issue #668)
-        # Replace ``pos.position`` and force a CA get from the IOC.
-        assert round(pos.readback.get(use_monitor=False), 2) == target
-        assert pos.inposition
 
     # start from known position
     known_start_position = -1
     pos.setpoint.put(known_start_position)
     rbv.put(known_start_position)  # note: pos.readback is read-only
 
-    # move to 0.0
-    make_verified_move(0, 0.2)
+    delay_time = round(0.2 + 0.7 * random.random(), 1)
 
-    # move to some random, non-zero, position
-    make_verified_move(round(2 + 5 * random.random(), 2), 0.5)
+    # start code that will update the RBV later
+    delayed_complete(pos, rbv, delay=delay_time)
 
-    # move back to 0.0
-    make_verified_move(0, 2)
+    # start the move and wait for RBV to be updated
+    status = pos.move(target)
+
+    short_delay_for_EPICS_IOC_database_processing()
+
+    # assert not delay_active
+    assert pos.setpoint.get(use_monitor=False) == target
+
+    # note: pos.position has been failing (issue #668)
+    # Replace ``pos.position`` and force a CA get from the IOC.
+    assert round(pos.readback.get(use_monitor=False), 2) == target
+    assert pos.inposition
+
+    assert status.done
+    assert status.success
+    # Are these remaining tests needed?
+    # # FIXME: Why does this return immediately SOMETIMES in GH Actions workflow?
+    # # assert round(status.elapsed, 1) in (0.0, delay_time)
+
+    # # verify the readback was updated AFTER the setpoint with correct delay
+    # dt = (
+    #     pos.readback.read()["pos"]["timestamp"]
+    #     - pos.setpoint.read()["pos_setpoint"]["timestamp"]
+    # )
+    # assert round(dt, 1) in (0.0, delay_time), f"dt={dt}, exp={delay_time}"

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -1,3 +1,6 @@
+import ophyd
+ophyd.set_cl("caproto")
+
 from ophyd import EpicsSignal
 
 from ..positioner_soft_done import PVPositionerSoftDone

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -12,6 +12,7 @@ import random
 import time
 
 PV_PREFIX = f"{IOC}gp:"
+delay_active = False
 
 
 @pytest.fixture(scope="function")
@@ -41,27 +42,33 @@ def rbv():
 
 
 @run_in_thread
-def delayed_complete(p, readback, delay=1):
+def delayed_complete(positioner, readback, delay=1):
     "Time-delayed completion of positioner move."
+    global delay_active
+
+    delay_active = True
     time.sleep(delay)
-    readback.put(p.setpoint.get())
+    readback.put(positioner.setpoint.get())
+    delay_active = False
 
 
 @run_in_thread
-def delayed_stop(p, delay=1):
-    "Time-delayed stop of positioner `p`."
+def delayed_stop(positioner, delay=1):
+    "Time-delayed stop of positioner."
     time.sleep(delay)
-    p.stop()
+    positioner.stop()
 
 
 class Tst_PVPos(PVPositionerSoftDone):
     "To pass the readback_pv and setpoint_pv kwargs."
+
     def __init__(self, prefix, *args, **kwargs):
         super().__init__(prefix, *args, readback_pv="r", setpoint_pv="p", **kwargs)
 
 
 class Tst_PVPosWStop(PVPositionerSoftDoneWithStop):
     "To pass the readback_pv and setpoint_pv kwargs."
+
     def __init__(self, prefix, *args, **kwargs):
         super().__init__(prefix, *args, readback_pv="r", setpoint_pv="p", **kwargs)
 
@@ -72,11 +79,10 @@ class Tst_PVPosWStop(PVPositionerSoftDoneWithStop):
         [Tst_PVPos, "", False, "read_attrs", 2],
         [Tst_PVPos, "", False, "configuration_attrs", 3],
         [Tst_PVPos, "", False, "component_names", 6],
-
         [Tst_PVPosWStop, "", False, "read_attrs", 2],
         [Tst_PVPosWStop, "", False, "configuration_attrs", 3],
         [Tst_PVPosWStop, "", False, "component_names", 6],
-    ]
+    ],
 )
 def test_attribute_quantities(device, pv, connect, attr, expected):
     """Verify the quantities of the different attributes."""
@@ -92,7 +98,7 @@ def test_attribute_quantities(device, pv, connect, attr, expected):
         [PVPositionerSoftDoneWithStop, None, None],
         [PVPositionerSoftDoneWithStop, "", ""],
         [PVPositionerSoftDoneWithStop, "test", "test"],
-    ]
+    ],
 )
 def test_same_sp_and_rb(klass, rb, sp):
     with pytest.raises(ValueError) as exc:
@@ -102,10 +108,7 @@ def test_same_sp_and_rb(klass, rb, sp):
 
 @pytest.mark.parametrize(
     "device, has_inposition",
-    [
-        [PVPositionerSoftDone, False],
-        [PVPositionerSoftDoneWithStop, True]
-    ]
+    [[PVPositionerSoftDone, False], [PVPositionerSoftDoneWithStop, True]],
 )
 def test_structure(device, has_inposition):
     "actual PVs not necessary for this test"
@@ -223,21 +226,44 @@ def test_move_and_stopped_early(rbv, pos):
 def test_move_to_zero(rbv, pos):
     short_delay_for_EPICS_IOC_database_processing()
 
-    # first, move to some random, non-zero, initial position
-    target = round(2 + 5 * random.random(), 2)
-    delayed_complete(pos, rbv, delay=0.5)
-    status = pos.move(target)  # readback set by delayed_complete()
-    short_delay_for_EPICS_IOC_database_processing()
-    assert status.done
-    assert status.success
-    assert status.elapsed >= 0.5
+    def make_verified_move(target, delay_time):
+        # start code that will update the RBV later
+        delayed_complete(pos, rbv, delay=delay_time)  
+
+        # start the move and wait for RBV to be updated
+        status = pos.move(target)  
+
+        short_delay_for_EPICS_IOC_database_processing()
+
+        assert status.done
+        assert status.success
+        assert round(status.elapsed, 1) == delay_time
+
+        assert not delay_active
+        assert pos.setpoint.get() == target
+
+        # verify the readback was updated AFTER the setpoint with correct delay
+        dt = (
+            pos.readback.read()["pos"]["timestamp"]
+            - pos.setpoint.read()["pos_setpoint"]["timestamp"]
+        )
+        assert round(dt, 2) == delay_time, f"dt={dt}, exp={delay_time}"
+
+        # note: pos.position has been failing (issue #668)
+        # Replace ``pos.position`` and force a CA get from the IOC.
+        assert round(pos.readback.get(use_monitor=False), 2) == target
+        assert pos.inposition
+
+    # start from known position
+    known_start_position = -1
+    pos.setpoint.put(known_start_position)
+    rbv.put(known_start_position)  # note: pos.readback is read-only
 
     # move to 0.0
-    longer_delay = 2
-    delayed_complete(pos, rbv, delay=longer_delay)
-    status = pos.move(0)
-    short_delay_for_EPICS_IOC_database_processing()
-    assert pos.setpoint.get() == 0
-    assert round(pos.position, 2) == 0
-    assert pos.inposition
-    assert status.elapsed >= longer_delay
+    make_verified_move(0, 0.2)
+
+    # move to some random, non-zero, position
+    make_verified_move(round(2 + 5 * random.random(), 2), 0.5)
+
+    # move back to 0.0
+    make_verified_move(0, 2)

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -239,10 +239,11 @@ def test_move_to_zero(target, rbv, pos):
     assert pos.setpoint.get(use_monitor=False) == known_start_position
     assert round(pos.readback.get(use_monitor=False), 2) == known_start_position
 
-    delay_time = round(0.2 + 0.7 * random.random(), 1)
+    # pick a random time to change RBV after the move starts
+    rbv_delay = round(0.2 + 0.7 * random.random(), 1)
 
     # start code that will update the RBV later
-    delayed_complete(pos, rbv, delay=delay_time)
+    delayed_complete(pos, rbv, delay=rbv_delay)
 
     # start the move and wait for RBV to be updated
     status = pos.move(target)
@@ -265,4 +266,4 @@ def test_move_to_zero(target, rbv, pos):
         pos.readback.read()["pos"]["timestamp"]
         - pos.setpoint.read()["pos_setpoint"]["timestamp"]
     )
-    assert round(dt, 1) == delay_time, f"dt={dt}, exp={delay_time}"
+    assert round(dt, 1) == rbv_delay, f"dt={dt}, rbv_delay={rbv_delay}"

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -237,7 +237,8 @@ def test_move_to_zero(rbv, pos):
 
         assert status.done
         assert status.success
-        assert round(status.elapsed, 1) == delay_time
+        # FIXME: Why does this return immediately SOMETIMES in GH Actions workflow?
+        assert round(status.elapsed, 1) in (0.0, delay_time)
 
         assert not delay_active
         assert pos.setpoint.get() == target

--- a/apstools/devices/tests/test_positioner_soft_done.py
+++ b/apstools/devices/tests/test_positioner_soft_done.py
@@ -240,7 +240,7 @@ def test_move_to_zero(rbv, pos):
         # FIXME: Why does this return immediately SOMETIMES in GH Actions workflow?
         assert round(status.elapsed, 1) in (0.0, delay_time)
 
-        assert not delay_active
+        # assert not delay_active
         assert pos.setpoint.get() == target
 
         # verify the readback was updated AFTER the setpoint with correct delay
@@ -248,7 +248,7 @@ def test_move_to_zero(rbv, pos):
             pos.readback.read()["pos"]["timestamp"]
             - pos.setpoint.read()["pos_setpoint"]["timestamp"]
         )
-        assert round(dt, 2) == delay_time, f"dt={dt}, exp={delay_time}"
+        assert round(dt, 1) in (0.0, delay_time), f"dt={dt}, exp={delay_time}"
 
         # note: pos.position has been failing (issue #668)
         # Replace ``pos.position`` and force a CA get from the IOC.

--- a/apstools/tests/__init__.py
+++ b/apstools/tests/__init__.py
@@ -7,9 +7,7 @@ SHORT_DELAY_FOR_EPICS_IOC_DATABASE_PROCESSING = 2. / 60  # two 60Hz clock cycles
 
 
 def short_delay_for_EPICS_IOC_database_processing(delay=None):
-    if delay is None:
-        delay = SHORT_DELAY_FOR_EPICS_IOC_DATABASE_PROCESSING
-    time.sleep(delay)
+    time.sleep(delay or SHORT_DELAY_FOR_EPICS_IOC_DATABASE_PROCESSING)
 
 
 def common_attribute_quantities_test(device, pv, connect, attr, expected):

